File: app/src/main/java/com/google/codelab/mlkit/MainActivity.java
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.codelab.mlkit;

import android.content.Context;
import android.content.res.AssetManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.util.Pair;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.codelab.mlkit.GraphicOverlay.Graphic;
import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.face.Face;
import com.google.mlkit.vision.face.FaceDetection;
import com.google.mlkit.vision.face.FaceDetector;
import com.google.mlkit.vision.face.FaceDetectorOptions;
import com.google.mlkit.vision.text.Text;
import com.google.mlkit.vision.text.TextRecognition;
import com.google.mlkit.vision.text.TextRecognizer;

import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

public class MainActivity extends AppCompatActivity implements AdapterView.OnItemSelectedListener {
    private ImageView mImageView;
    private Button mTextButton;
    private Bitmap mSelectedImage;
    private GraphicOverlay mGraphicOverlay;
    private Integer mImageMaxWidth;
    private Integer mImageMaxHeight;
    private TextView mtextView;


    //lista de marcas a serem buscadas
    List<String> listaDeMarcas = Arrays.asList(
            "ama", "ambev", "brahma", "budweiser", "stella", "artois", "beck's", "becks", "beck",
            "skol", "wälls", "walls", "xwalls", "x-walls", "spaten", "bohemia", "colorado",
            "corona", "beats", "antartica"
    );


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mImageView = findViewById(R.id.image_view);

        mTextButton = findViewById(R.id.button_text);


        mGraphicOverlay = findViewById(R.id.graphic_overlay);
        mTextButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                runTextRecognition();
            }
        });
        Spinner dropdown = findViewById(R.id.spinner);
        String[] items = new String[]{"Spaten", "Diferentes", "Beer", "Budweiser", "Bud Poster",
                "Ambev", "Marcas", "Ambev Ama", "Becks Corona", "Bohemia Poster", "Brahma Skol",
                "Original", "Skol Brahma", "Skol Mao"};

        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout
                .simple_spinner_dropdown_item, items);
        dropdown.setAdapter(adapter);
        dropdown.setOnItemSelectedListener(this);

        mtextView = findViewById(R.id.text_view);

    }

    private void runTextRecognition() {
        InputImage image = InputImage.fromBitmap(mSelectedImage, 0);
        TextRecognizer recognizer = TextRecognition.getClient();
        mTextButton.setEnabled(false);
        recognizer.process(image)
                .addOnSuccessListener(
                        new OnSuccessListener<Text>() {
                            @Override
                            public void onSuccess(Text texts) {
                                mTextButton.setEnabled(true);
                                processTextRecognitionResult(texts);
                            }
                        })
                .addOnFailureListener(
                        new OnFailureListener() {
                            @Override
                            public void onFailure(@NonNull Exception e) {
                                // Task failed with an exception
                                mTextButton.setEnabled(true);
                                e.printStackTrace();
                            }
                        });
    }

    private void processTextRecognitionResult(Text texts) {
        mtextView.setText(" ");
        List<Text.TextBlock> blocks = texts.getTextBlocks();
        if (blocks.size() == 0) {
            showToast("No text found");
            return;
        }
        mGraphicOverlay.clear();
        for (int i = 0; i < blocks.size(); i++) {
            List<Text.Line> lines = blocks.get(i).getLines();
            for (int j = 0; j < lines.size(); j++) {
                List<Text.Element> elements = lines.get(j).getElements();
                showPossibleMatches(elements);
                for (int k = 0; k < elements.size(); k++) {
                    if (listaDeMarcas.contains(elements.get(k).getText().toLowerCase())) {
                        Graphic textGraphic = new TextGraphic(mGraphicOverlay, elements.get(k));
                        mGraphicOverlay.add(textGraphic);
                    }
                }
            }
        }
    }

    private void showToast(String message) {
        Toast.makeText(getApplicationContext(), message, Toast.LENGTH_SHORT).show();
    }

    // Functions for loading images from app assets.

    // Returns max image width, always for portrait mode. Caller needs to swap width / height for
    // landscape mode.
    private Integer getImageMaxWidth() {
        if (mImageMaxWidth == null) {
            // Calculate the max width in portrait mode. This is done lazily since we need to
            // wait for
            // a UI layout pass to get the right values. So delay it to first time image
            // rendering time.
            mImageMaxWidth = mImageView.getWidth();
        }

        return mImageMaxWidth;
    }

    // Returns max image height, always for portrait mode. Caller needs to swap width / height for
    // landscape mode.
    private Integer getImageMaxHeight() {
        if (mImageMaxHeight == null) {
            // Calculate the max width in portrait mode. This is done lazily since we need to
            // wait for
            // a UI layout pass to get the right values. So delay it to first time image
            // rendering time.
            mImageMaxHeight =
                    mImageView.getHeight();
        }

        return mImageMaxHeight;
    }

    // Gets the targeted width / height.
    private Pair<Integer, Integer> getTargetedWidthHeight() {
        int targetWidth;
        int targetHeight;
        int maxWidthForPortraitMode = getImageMaxWidth();
        int maxHeightForPortraitMode = getImageMaxHeight();
        targetWidth = maxWidthForPortraitMode;
        targetHeight = maxHeightForPortraitMode;
        return new Pair<>(targetWidth, targetHeight);
    }

    public void onItemSelected(AdapterView<?> parent, View v, int position, long id) {
        mGraphicOverlay.clear();
        switch (position) {
            case 0:
                mSelectedImage = getBitmapFromAsset(this, "spaten.jpg");
                break;
            case 1:
                // Whatever you want to happen when the thrid item gets selected
                mSelectedImage = getBitmapFromAsset(this, "diferentes.jpg");
                break;
            case 2:
                mSelectedImage = getBitmapFromAsset(this, "beer.jpg");
                break;
            case 3:
                mSelectedImage = getBitmapFromAsset(this, "budweiser.jpg");
                break;
            case 4:
                mSelectedImage = getBitmapFromAsset(this, "bud_posrter.jpg");
                break;
            case 5:
                mSelectedImage = getBitmapFromAsset(this, "ambev.jpg");
                break;
            case 6:
                mSelectedImage = getBitmapFromAsset(this, "marcas.jpg");
                break;
            case 7:
                mSelectedImage = getBitmapFromAsset(this, "ambev-ama.jpg");
                break;
            case 8:
                mSelectedImage = getBitmapFromAsset(this, "becks_corona.jpg");
                break;
            case 9:
                mSelectedImage = getBitmapFromAsset(this, "bohemia_poster.jpg");
                break;
            case 10:
                mSelectedImage = getBitmapFromAsset(this, "brahma_skol.jpg");
                break;
            case 11:
                mSelectedImage = getBitmapFromAsset(this, "original.jpg");
                break;
            case 12:
                mSelectedImage = getBitmapFromAsset(this, "skol_brahma.jpg");
                break;
            case 13:
                mSelectedImage = getBitmapFromAsset(this, "skol_mao.jpg");
                break;
        }
        if (mSelectedImage != null) {
            // Get the dimensions of the View
            Pair<Integer, Integer> targetedSize = getTargetedWidthHeight();

            int targetWidth = targetedSize.first;
            int maxHeight = targetedSize.second;

            // Determine how much to scale down the image
            float scaleFactor =
                    Math.max(
                            (float) mSelectedImage.getWidth() / (float) targetWidth,
                            (float) mSelectedImage.getHeight() / (float) maxHeight);

            Bitmap resizedBitmap =
                    Bitmap.createScaledBitmap(
                            mSelectedImage,
                            (int) (mSelectedImage.getWidth() / scaleFactor),
                            (int) (mSelectedImage.getHeight() / scaleFactor),
                            true);

            mImageView.setImageBitmap(resizedBitmap);
            mSelectedImage = resizedBitmap;
        }
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {
        // Do nothing
    }

    public static Bitmap getBitmapFromAsset(Context context, String filePath) {
        AssetManager assetManager = context.getAssets();

        InputStream is;
        Bitmap bitmap = null;
        try {
            is = assetManager.open(filePath);
            bitmap = BitmapFactory.decodeStream(is);
        } catch (IOException e) {
            e.printStackTrace();
        }

        return bitmap;
    }

    private void showPossibleMatches(List<Text.Element> elements) {
        for (Text.Element reconhecido : elements) {
            for (String marcas : listaDeMarcas) {
                String identificado = reconhecido.getText().toLowerCase();
                if (identificado.length() >= 3 && !marcas.equals(identificado) && marcas.startsWith(identificado.substring(0, 3)))
                    mtextView.append("Marca: " + marcas + "->" + reconhecido.getText() + "\n");

            }

        }
    }
}

File: app/src/main/java/com/lentimosystems/mobivisionkt/MainActivity.kt
package com.lentimosystems.mobivisionkt

import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.provider.MediaStore
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.core.content.ContextCompat
import com.google.android.gms.tasks.Task
import com.google.firebase.ml.vision.FirebaseVision
import com.google.firebase.ml.vision.common.FirebaseVisionImage
import com.google.firebase.ml.vision.text.FirebaseVisionText
import java.util.jar.Manifest

class MainActivity : AppCompatActivity() {
    private lateinit var btn_image: Button
    private lateinit var imageView: ImageView
    private lateinit var txt_image: TextView
    private val listaDeMarcas = listOf(
        "ama", "ambev", "brahma", "budweiser", "stella", "artois", "beck's", "becks", "beck",
        "skol", "wälls", "walls", "xwalls", "x-walls", "spaten", "bohemia", "colorado",
        "corona", "beats", "antartica", "serramalte"
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        btn_image = findViewById(R.id.btnImage)
        imageView = findViewById(R.id.image)
        txt_image = findViewById(R.id.txt_image)

        btn_image.setOnClickListener { view: View? ->
            chooseImage()
        }

    }

    private fun chooseImage() {
        if (ContextCompat.checkSelfPermission(
                this,
                android.Manifest.permission.CAMERA
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            requestPermissions(arrayOf(android.Manifest.permission.CAMERA), 123)
        }
        val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        startActivityForResult(intent, 123)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        //Extract the image
        val bundle = data!!.extras
        val bitmap = bundle!!["data"] as Bitmap?
        imageView.setImageBitmap(bitmap)

        //Create a FirebaseVisionImage object from your image/bitmap.
        val firebaseVisionImage = FirebaseVisionImage.fromBitmap(bitmap!!)
        val firebaseVision = FirebaseVision.getInstance()
        val firebaseVisionTextRecognizer = firebaseVision.onDeviceTextRecognizer

        //Process the Image
        val task = firebaseVisionTextRecognizer.processImage(firebaseVisionImage)
        task.addOnSuccessListener { firebaseVisionText: FirebaseVisionText ->
            //Set recognized text from image in our TextView
            //Processar apenas para as marcas: filtro
            val text = firebaseVisionText.text
            txt_image.text = " "
            Toast.makeText(
                this,
                "Processando",
                Toast.LENGTH_SHORT
            ).show()
            text.split("\n").forEach {
                showMatches(it)
            }
        }
        task.addOnFailureListener { e: Exception ->
            Toast.makeText(
                this,
                e.message,
                Toast.LENGTH_SHORT
            ).show()
        }
    }

    private fun showMatches(blockText: String) {
        val identificado = blockText.toLowerCase()

        if (listaDeMarcas.contains(identificado))
            txt_image.append(blockText + "\n")
        else {
            listaDeMarcas.forEach { marca ->
                if (identificado.count() >= 3 && marca.startsWith(identificado.substring(0, 3)))
                    txt_image.append("Marca: $marca->$identificado\n")
            }
        }
    }
}